homeassistant/components/device_tracker/huawei_hilink.py

```diff
@@ -0,0 +1,217 @@
+        self.username = config[CONF_USERNAME]
+        self.password = config[CONF_PASSWORD]
+
+        self.api_location = 'http://{}/api'.format(self.host)
+        self.auth_info = AuthInfo(session_id='wrong',
+                                  verification_token='wrong')
+
+        self.last_results = []
+        _LOGGER.info("Scanner initialized")
+
+    def scan_devices(self):
+        """Scan for new devices."""
+        self.update_info()
+        return [device.mac for device in self.last_results]
+
+    def get_device_name(self, mac):
+        """Get device name from mac."""
+        names = [device.name for device in self.last_results
+                 if device.mac == mac]
+        if names:
+            return names[0]
+        return None
+
+    def update_info(self):
+        """Ensure the information from the router is up to date.
+
+        Retry once if fetching devices fails with fresh login. Return True if
+        scanning successful.
+        """
+        _LOGGER.info("Scanning...")
+        devices = _fetch_devices(self.api_location, self.auth_info)
+        if devices is None:
+            _LOGGER.info("Obtaining devices failed. Try to login")
+            success_login = self._login()
+            if not success_login:
+                return False
+            devices = _fetch_devices(self.api_location, self.auth_info)
+            if devices is None:
+                _LOGGER.warning("Fetching devices failed again. Stop")
+                return False
+
+        self.last_results = devices
+        return True
+
+    def _login(self):
+        """Login and update authorization info.
+
+        Retry once with fresh authorization info. Return True if login
+        successful.
+        """
+        self.auth_info = _login(self.api_location,
+                                self.username,
+                                self.password,
+                                self.auth_info)
+        if self.auth_info is None:
+            _LOGGER.info("Logging in failed. Try to obtain auth info again")
+            self._obtain_auth_info()
+            self.auth_info = _login(self.api_location,
+                                    self.username,
+                                    self.password,
+                                    self.auth_info)
+            if self.auth_info is None:
+                _LOGGER.warning("Logging in failed with fresh auth info. Stop")
+                return False
+
+        return True
+
+    def _obtain_auth_info(self):
+        """Obtain and update authorization info."""
+        self.auth_info = _obtain_auth_info(self.api_location)
+
+
+def _obtain_auth_info(api_location):
+    """Return authorization info with session id and verification token.
+
+    Return None if operation not successful.
+    """
+    import defusedxml.ElementTree as ET
+
+    _LOGGER.info("Obtaining session id and verification token")
+
+    response = requests.get('{}/webserver/SesTokInfo'.format(api_location))
```

Comment:
This is too much protocol specific logic. Please extract as a third party lib to interact with Huawei routers.